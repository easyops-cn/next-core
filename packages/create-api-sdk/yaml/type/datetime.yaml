_version_: 2.0
_kind_: type

name: datetime
type: string
description: 日期时间（支持ISO 8601规范）

validate:
  pattern: /^((?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[0-1]|0[1-9]|[1-2][0-9])[T ](2[0-3]|[0-1][0-9]):([0-5][0-9]):[0-5][0-9](\.[0-9]+)?(Z|[+-](?:2[0-3]|[0-1][0-9]):[0-5][0-9])?$/
