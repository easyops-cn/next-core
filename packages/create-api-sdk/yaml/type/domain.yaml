_version_: 2.0
_kind_: type

name: domain
type: string
description: 域名

validate:
  pattern: /(?:[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?\.)+[a-z0-9][a-z0-9-]{0,61}[a-z0-9]/
