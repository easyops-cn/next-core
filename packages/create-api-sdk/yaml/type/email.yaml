_version_: 2.0
_kind_: type

name: email
type: string
description: 电子邮箱地址

validate:
  pattern: /^([A-Za-z0-9_\-\.\u4e00-\u9fa5])+\@([A-Za-z0-9_\-\.])+\.([A-Za-z]{2,8})$/
