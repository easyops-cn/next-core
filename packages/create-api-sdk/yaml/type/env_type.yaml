_version_: 2.0
_kind_: type

name: env_type
type: int
enum: [1, 3, 7, 15]
description: 版本类型 1 开发, 3 测试, 7 预发布, 15 生产
