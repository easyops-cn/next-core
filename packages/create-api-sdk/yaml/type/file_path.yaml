_version_: 2.0
_kind_: type

name: file_path
type: string
description: 文件路径（兼容linux和windows）

validate:
  pattern: /^((\/[^\/:\*\?""<>\|\r\n]+)+)|(\/)|[a-zA-Z]:(\\[^\\:\*\?""<>\|\r\n]+)*$/
