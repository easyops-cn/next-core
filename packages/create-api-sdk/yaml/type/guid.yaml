_version_: 2.0
_kind_: type

name: guid
type: string
description: 全局唯一ID

validate:
  pattern: /^[?a-zA-Z0-9]{8}-[?a-zA-Z0-9]{4}-[?a-zA-Z0-9]{4}-[?a-zA-Z0-9]{4}-[?a-zA-Z0-9]{12}$/
