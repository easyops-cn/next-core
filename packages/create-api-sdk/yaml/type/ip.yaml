_version_: 2.0
_kind_: type

name: ip
type: string
description: 网络地址（包括IPV4和IPV6）

validate:
  pattern: /^((2(5[0-5]{1}|[0-4]\d{1})|[0-1]?\d{1,2})(\.(2(5[0-5]{1}|[0-4]\d{1})|[0-1]?\d{1,2})){3})|\s*((([0-9A-Fa-f]{1,4}:){7}([0-9A-Fa-f]{1,4}|:))|(([0-9A-Fa-f]{1,4}:){6}(:[0-9A-Fa-f]{1,4}|((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){5}(((:[0-9A-Fa-f]{1,4}){1,2})|:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3})|:))|(([0-9A-Fa-f]{1,4}:){4}(((:[0-9A-Fa-f]{1,4}){1,3})|((:[0-9A-Fa-f]{1,4})?:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){3}(((:[0-9A-Fa-f]{1,4}){1,4})|((:[0-9A-Fa-f]{1,4}){0,2}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){2}(((:[0-9A-Fa-f]{1,4}){1,5})|((:[0-9A-Fa-f]{1,4}){0,3}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(([0-9A-Fa-f]{1,4}:){1}(((:[0-9A-Fa-f]{1,4}){1,6})|((:[0-9A-Fa-f]{1,4}){0,4}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:))|(:(((:[0-9A-Fa-f]{1,4}){1,7})|((:[0-9A-Fa-f]{1,4}){0,5}:((25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)(\.(25[0-5]|2[0-4]\d|1\d\d|[1-9]?\d)){3}))|:)))(%.+)?\s*$/
