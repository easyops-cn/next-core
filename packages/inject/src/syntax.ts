import { PipeCall } from "@next-core/pipes";
import { tokenize } from "./lexical.js";
import {
  Token,
  TokenType,
  InjectableString,
  Placeholder,
} from "./interfaces.js";

export function parseInjectableString(
  raw: string,
  symbols: string | string[]
): InjectableString {
  return parseTokens(tokenize(raw, symbols));
}

function parseTokens(tokens: Token[]): InjectableString {
  const tree: InjectableString = {
    type: "InjectableString",
    elements: [],
  };

  let token: Token | undefined;

  function optionalToken(type: TokenType): boolean {
    if (type === tokens[0].type) {
      token = tokens.shift();
      return true;
    }
    return false;
  }

  function acceptToken(type: TokenType | TokenType[]): void {
    token = tokens.shift() as Token;
    if (
      Array.isArray(type) ? !type.includes(token.type) : type !== token.type
    ) {
      throw new Error(`Expected token: ${type}, received token: ${token.type}`);
    }
  }

  while (tokens.length > 0) {
    if (optionalToken(TokenType.Raw)) {
      tree.elements.push({
        type: "RawString",
        value: token.value,
      });
    } else {
      acceptToken(TokenType.BeginPlaceHolder);
      const start = token.loc.start;
      const symbol = token.value;
      acceptToken(TokenType.Field);

      const placeholder: Placeholder = {
        type: "Placeholder",
        symbol,
        field: token.value,
        defaultValue: undefined,
        pipes: [],
        loc: {
          start,
          end: start,
        },
      };
      tree.elements.push(placeholder);

      if (optionalToken(TokenType.BeginDefault)) {
        acceptToken([TokenType.JsonValue, TokenType.LiteralString]);
        placeholder.defaultValue = token.value;
      }

      while (optionalToken(TokenType.BeginPipe)) {
        acceptToken(TokenType.PipeIdentifier);
        const pipe: PipeCall = {
          type: "PipeCall",
          identifier: token.value,
          parameters: [],
        };
        placeholder.pipes.push(pipe);

        while (optionalToken(TokenType.BeginPipeParameter)) {
          acceptToken([TokenType.JsonValue, TokenType.LiteralString]);
          pipe.parameters.push(token.value);
        }
      }

      acceptToken(TokenType.EndPlaceHolder);
      placeholder.loc.end = token.loc.end;
    }
  }

  return tree;
}
