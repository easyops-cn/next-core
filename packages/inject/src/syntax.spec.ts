import { parseInjectableString } from "./syntax.js";
import type { InjectableString } from "./interfaces.js";

describe("parseInjectableString", () => {
  it.each<[string, InjectableString, string | string[]]>([
    [
      "",
      {
        type: "InjectableString",
        elements: [],
      },
      "$",
    ],
    [
      "good",
      {
        type: "InjectableString",
        elements: [
          {
            type: "RawString",
            value: "good",
          },
        ],
      },
      "$",
    ],
    [
      "${{good}}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "RawString",
            value: "${{good}}",
          },
        ],
      },
      "$",
    ],
    [
      "#{abc}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "RawString",
            value: "#{abc}",
          },
        ],
      },
      "$",
    ],
    [
      "${}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "Placeholder",
            symbol: "$",
            field: "",
            defaultValue: undefined,
            pipes: [],
            loc: {
              start: 0,
              end: 3,
            },
          },
        ],
      },
      "$",
    ],
    [
      "${QUERY.page}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "Placeholder",
            symbol: "$",
            field: "QUERY.page",
            defaultValue: undefined,
            pipes: [],
            loc: {
              start: 0,
              end: 13,
            },
          },
        ],
      },
      "$",
    ],
    [
      "${QUERY.page=}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "Placeholder",
            symbol: "$",
            field: "QUERY.page",
            defaultValue: "",
            pipes: [],
            loc: {
              start: 0,
              end: 14,
            },
          },
        ],
      },
      "$",
    ],
    [
      "${QUERY.page=-1}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "Placeholder",
            symbol: "$",
            field: "QUERY.page",
            defaultValue: -1,
            pipes: [],
            loc: {
              start: 0,
              end: 16,
            },
          },
        ],
      },
      "$",
    ],
    [
      "asc=${QUERY.asc=true|number}&q=@{quality=good|split:-}",
      {
        type: "InjectableString",
        elements: [
          {
            type: "RawString",
            value: "asc=",
          },
          {
            type: "Placeholder",
            symbol: "$",
            field: "QUERY.asc",
            defaultValue: true,
            pipes: [
              {
                type: "PipeCall",
                identifier: "number",
                parameters: [],
              },
            ],
            loc: {
              start: 4,
              end: 28,
            },
          },
          {
            type: "RawString",
            value: "&q=",
          },
          {
            type: "Placeholder",
            symbol: "@",
            field: "quality",
            defaultValue: "good",
            pipes: [
              {
                type: "PipeCall",
                identifier: "split",
                parameters: ["-"],
              },
            ],
            loc: {
              start: 31,
              end: 54,
            },
          },
        ],
      },
      ["@", "$"],
    ],
    [
      '${ some.field[0].path = ["complex","value\\n"] | map : {"a":-12.34E+5} : true | join : }',
      {
        type: "InjectableString",
        elements: [
          {
            type: "Placeholder",
            symbol: "$",
            field: "some.field[0].path",
            defaultValue: ["complex", "value\n"],
            pipes: [
              {
                type: "PipeCall",
                identifier: "map",
                parameters: [{ a: -12.34e5 }, true],
              },
              {
                type: "PipeCall",
                identifier: "join",
                parameters: [""],
              },
            ],
            loc: {
              start: 0,
              end: 87,
            },
          },
        ],
      },
      "$",
    ],
  ])("parseInjectableString should work for %j", (raw, tree, symbols) => {
    expect(parseInjectableString(raw, symbols)).toEqual(tree);
  });

  it.each<string>(["${a|b=c}", "${a=[}", "${a|0}", "${a=[{}}{]}"])(
    "should throw when parsing %j",
    (raw) => {
      expect(() => {
        parseInjectableString(raw, "$");
      }).toThrowErrorMatchingSnapshot();
    }
  );
});
