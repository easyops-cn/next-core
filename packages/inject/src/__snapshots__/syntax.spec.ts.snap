// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseInjectableString should throw when parsing "\${a=[{}}{]}" 1`] = `"Unexpected token } in JSON at position 3"`;

exports[`parseInjectableString should throw when parsing "\${a=[}" 1`] = `"Failed to match a JSON value at index 4 near: \\"[}\\""`;

exports[`parseInjectableString should throw when parsing "\${a|0}" 1`] = `"Expected a pipe identifier at index 4 near: \\"0}\\""`;

exports[`parseInjectableString should throw when parsing "\${a|b=c}" 1`] = `"Expected a placeholder end '}' at index 5 near: \\"=c}\\""`;
